lib/twirp/protoc_plugin/descriptor_ext/service_descriptor_proto_ext.rb

```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+require_relative "../../../core_ext/string/camel_case"
+
+class Google::Protobuf::ServiceDescriptorProto
+  def service_class_name
+    # The generated service class name should end in "Service"; But only append the
+    # suffix if the service is not already well-named.
+    service_class_name = if name.end_with?("Service")
+      name
+    else
+      name + "Service"
+    end
+    service_class_name.camel_case
+  end
+
+  def client_class_name
+    # Strip the "Service" suffix if present for better readability.
+    (name.delete_suffix("Service") + "Client").camel_case
```

Comment:
Is this clearer?

```suggestion
  def base_class_name
    name.delete_suffix("Service")
  end
  
  def service_class_name
    # The generated service class name should end in "Service"; 
    # A well-named service may already end with "Service" but we can't
    # guarantee it, so use base_class_name to avoid "ServiceService"
    (base_class_name + "Service").camel_case
  end

  def client_class_name
    # Strip the "Service" suffix if present for better readability.
    (base_class_name + "Client").camel_case
```

Reply:
Yeah, I like where you're going with that. Not sure I like `base_class_name` though given `base` connotations around inheritance. Should also move that to `private`. How about `name_without_suffix`?

Reply:
I like both of those suggestions.

Reply:
Eh, in the common case, this ends up deleting "Service" and then adding it back. The original has the conditional there but does less work.

Reply:
micro optimization? Still has to inspect the string to find out if it needs to do the work.

Reply:
and then you delete it in the _client method anyway. Go for readability.

Reply:
heck, if you're micro-optimizing, move `.camel_case` into `name_without_suffix`